project-base/app/src/Model/Product/ProductRepository.php

```diff
@@ -371,13 +371,15 @@ protected function applyOrdering(
     public function getProductIdsByCatnums(array $catnums): array
     {
         $result = $this->em->createQueryBuilder()
-            ->select('p.id')
+            ->select('p.id, p.catnum')
             ->from(Product::class, 'p')
             ->where('p.catnum IN (:catnums)')
             ->setParameter('catnums', $catnums)
             ->getQuery()
             ->getScalarResult();
 
-        return array_column($result, 'id');
+        $catnumToIdMap = array_column($result, 'id', 'catnum');
+
+        return array_values(array_filter(array_map(static fn (string $catnum) => $catnumToIdMap[$catnum] ?? null, $catnums)));
```

Comment:
This is really badly readable, can you refactor it?

Reply:
b6194ee40 - is it better now? :slightly_smiling_face:

Reply:
...and second iteration of refactoring :smile: 2865179aa